File: password_gen/pass_gen.py
import random
import string
import os

def generate_password():
    # Define criteria
    length = random.randint(9, 14)  # Random length between 9 and 14 characters
    uppercase = random.choice(string.ascii_uppercase)  # At least one uppercase
    number = random.choice(string.digits)  # At least one number
    special_char = random.choice('@#$%&*')  # At least one of the specified special characters

    # Generate remaining characters
    remaining_chars = string.ascii_letters + string.digits + '@#$%&*'
    password = ''.join(random.choice(remaining_chars) for _ in range(length - 3))

    # Shuffle and insert the required characters at random positions
    password_list = list(password)
    password_list.insert(random.randint(0, length - 1), uppercase)
    password_list.insert(random.randint(0, length), number)
    password_list.insert(random.randint(0, length + 1), special_char)

    # Convert the list back to a string
    final_password = ''.join(password_list)

    return final_password

def save_to_log(password):
    log_file_path = os.path.join(os.path.dirname(__file__), 'pass_log.txt')
    with open(log_file_path, 'a') as log_file:
        log_file.write(password + '\n')

# Example usage
password = generate_password()
print("Generated Password:", password)

# Save the password to the log file
save_to_log(password)

File: password_gen/passgen.rs
use rand::{thread_rng, Rng};
use std::fs::OpenOptions;
use std::io::Write;
use std::path::Path;

fn generate_password() -> String {
    // Define criteria
    let length = thread_rng().gen_range(9..=14); // Random length between 9 and 14 characters
    let uppercase = char::from(thread_rng().gen_range(b'A'..=b'Z')); // At least one uppercase
    let number = char::from(thread_rng().gen_range(b'0'..=b'9')); // At least one number
    let special_char = ['@', '#', '$', '%', '&', '*']
        .choose(&mut thread_rng())
        .unwrap(); // At least one of the specified special characters

    // Generate remaining characters
    let remaining_chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789@#$%&*";
    let mut password: String = thread_rng()
        .sample_iter(rand::distributions::Alphanumeric)
        .take(length - 3)
        .collect();

    // Shuffle and insert the required characters at random positions
    let position_uppercase = thread_rng().gen_range(0..=length - 1);
    let position_number = thread_rng().gen_range(0..=length);
    let position_special_char = thread_rng().gen_range(0..=length + 1);

    password.insert(position_uppercase, uppercase);
    password.insert(position_number, number);
    password.insert(position_special_char, *special_char);

    password
}

fn save_to_log(password: &str) {
    let log_file_path = Path::new(env!("CARGO_MANIFEST_DIR")).join("pass_log.txt");
    let mut log_file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(log_file_path)
        .expect("Unable to open log file");

    writeln!(log_file, "{}", password).expect("Unable to write to log file");
}

fn main() {
    // Example usage
    let password = generate_password();
    println!("Generated Password: {}", password);

    // Save the password to the log file
    save_to_log(&password);
}
